### PR TITLE
bug 1060695 - add a frame script for selection events

### PR DESCRIPTION
**Pull request by zombie**

### CODE REVIEW

File: lib/framescript/selection-events.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+function observer(subject, topic) {
+  if (!docShell)
+    obsSvc.removeObserver(observer, topic);
+  else {
+    let window = subject.defaultView;
+    if (!window || window.top !== content)
+      return;
+
+    let selection = window.getSelection();
+    if (!selection || selections.get(window) === selection)
```
Comment (Mossop) [inline lib/framescript/selection-events.js]: When is selection ever null?

File: lib/framescript/selection-events.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+obsSvc.addObserver(observer, 'document-shown', false);
+obsSvc.addObserver(observer, 'document-element-inserted', false);
+
+function observer(subject, topic) {
+  if (!docShell)
+    obsSvc.removeObserver(observer, topic);
+  else {
+    let window = subject.defaultView;
+    if (!window || window.top !== content)
```
Comment (Mossop) [inline lib/framescript/selection-events.js]: When is window ever null?

Comment (Mossop) [inline lib/framescript/selection-events.js]: Please add a comment to the effect that we get observer notifications for every frame in the process so we're filtering to only the ones for this frame.

File: lib/framescript/selection-events.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+const INTERACTIVE = Ci.nsISelectionListener.MOUSEUP_REASON |
+                    Ci.nsISelectionListener.KEYPRESS_REASON |
+                    Ci.nsISelectionListener.SELECTALL_REASON;
+
+let selections = new WeakMap();
+let isPrivate = docShell.QueryInterface(Ci.nsILoadContext).usePrivateBrowsing;
+
+obsSvc.addObserver(observer, 'document-shown', false);
+obsSvc.addObserver(observer, 'document-element-inserted', false);
```
Comment (Mossop) [inline lib/framescript/selection-events.js]: Might be worth testing if using the DOM event "DOMWindowCreated" is better than using global observer notifications.

**Review by zombie — COMMENTED** [review]
Looks great, thanks!

 [blind-review](https://github.com/zombie/blind-reviews 'Review performed in blind mode')

**Pull request closed without merging**